### PR TITLE
Check for the number of target qubits

### PR DESCRIPTION
**Pull request by AlexandreF-1qbit**
Naive attempt at fixing https://github.com/goodchemistryco/Tangelo/issues/149. It checks the number of target qubits for some gates. It does not prevent multicontrolled-qubit gates nor custom gates.

### CODE REVIEW

**Review by JamesB-1qbit — APPROVED** [review]

File: tangelo/linq/gate.py [hunk 1]
```diff
@@ -84,6 +89,18 @@ def check_qubit_indices(qubit_list, label):
...
+        # the X_QUBIT_GATES sets), ignore the check. We also do not prevent
+        # the creation of multi-controlled gates (begins with "C").
+        if name.upper() in ONE_QUBIT_TARGET_GATES:
+            n_qubits = 1
+        elif name.upper() in TWO_QUBIT_TARGET_GATES:
+            n_qubits = 2
+        else:
+            n_qubits = len(target)
+        if len(target) != n_qubits:
+            raise ValueError(f"There are too many target/control qubits for this gate.")
```
Comment (ValentinS4t1qbit) [inline tangelo/linq/gate.py:102]: 1. Replace "this gate" by the actual name of the gate, let's be useful there. I'd rather use `n_targets` than `n_qubits` as we use `n_qubits` many other places. We could say something like.

`f"Gate {g.name} : expected {n_qubits} target qubits, but got {target}"`

2. Your current code only seems to check that the number of targets is good, I'm not sure why your current error message says 'control' as well. Should there be a check for control, or none at all ?

File: tangelo/linq/gate.py [hunk 2]
```diff
@@ -24,6 +24,11 @@
 ONE_QUBIT_GATES = {"H", "X", "Y", "Z", "S", "T", "RX", "RY", "RZ", "PHASE"}
 TWO_QUBIT_GATES = {"CNOT", "CX", "CY", "CZ", "CRX", "CRY", "CRZ", "CPHASE", "XX", "SWAP"}
 THREE_QUBIT_GATES = {"CSWAP"}
+
+ONE_QUBIT_TARGET_GATES = {"H", "X", "Y", "Z", "S", "T", "RX", "RY", "RZ", "PHASE",
```
Comment (ValentinS4t1qbit) [inline tangelo/linq/gate.py:28]: I suggest `ONE_TARGET_GATES` and `TWO_TARGET_GATES`.

File: tangelo/linq/tests/test_gates.py [hunk 3]
```diff
@@ -94,6 +94,15 @@ def test_gate_equality_modulo_twopi(self):
         self.assertTrue(g1 == g3)
         self.assertTrue(g2 == g3)
 
+    def test_too_many_qubits_on_gates(self):
+        """ Test the behavior when too many qubits are selected for a gate. """
+
+        # Trying to create a Hadamard gate acting on qubits 0 and 1.
```
Comment (ValentinS4t1qbit) [inline tangelo/linq/tests/test_gates.py:100]: `Try`. Do not conjugate verbs. Same with line 103

**Review by ValentinS4t1qbit — CHANGES_REQUESTED** [review]
Thank you for addressing this long-lasting bug, which gives us the opportunity to show on Github that we follow up on those !

File: tangelo/linq/gate.py (hunk 1 above)
Comment (AlexandreF-1qbit) [reply tangelo/linq/gate.py:102]: Oh this is a mistake as I changed my mind last minute on the control check.

**Review by ValentinS4t1qbit — APPROVED** [review]

**Review by ValentinS4t1qbit — APPROVED** [review]

**Pull request merged**